### PR TITLE
Added identify api and changed a lot of stuff

### PR DESCRIPTION
**Pull request by lunfel**
List of changes:

- Added the possibility to call the identify api with existing UserProperties (useful when updating UserProperties without sending events)
- Changed to the http v2 api endpoint. This should close issue #25.
- Instead of logging errors silently, errors are now thrown. This should close issue #16.
- Added the composer.lock file which is a [good practice to commit in the repository](https://getcomposer.org/doc/01-basic-usage.md#commit-your-composer-lock-file-to-version-control)
- Updated tests to cover as most as I could of the new code
- Updated phpunit to version 9.x

This pull request adds breaking change to the api of this repository. A major version change will be required to merge this in.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -17,10 +17,12 @@
     ],
     "require": {
         "php": ">=7.2",
```
Comment (jonyo) [inline composer.json:19]: Need to update this to `>=7.3` to match test coverage.

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,6 @@
 composer.phar
-composer.lock
```
Comment (jonyo) [inline .gitignore:2]: While it is recommended to commit composer.lock on your projects, it is optional for library packages, and it is actually ignored when the library is used inside another package as a requirement.

See https://getcomposer.org/doc/02-libraries.md#lock-file

We prefer not to lock it so that any new PR is tested against the latest versions (within versioning requirements set in composer.json) so it would catch any potential problems with newer releases.  And if it is because a newer version is not compatible, then the `composer.js` file should be updated to reflect that.

File: composer.json [hunk 3]
```diff
@@ -17,10 +17,12 @@
...
-        "psr/log": "^1.0"
+        "psr/log": "^1.0",
+        "ext-curl": "*",
+        "ext-json": "*"
     },
     "require-dev": {
-        "phpunit/phpunit": "8.3.*",
+        "phpunit/phpunit": "^9.0",
```
Comment (jonyo) [inline composer.json:25]: 👍

File: src/Amplitude.php [hunk 4]
```diff
@@ -179,39 +154,34 @@ public function resetQueue()
...
-     *
-     * You can also pass in an event or array of event properties.  If you pass in an event, it will be set as the
-     * event to be used for the next call to queueEvent() or logEvent()
-     *
-     * @param null|array|\Zumba\Amplitude\Event Can pass in an event to set as the next event to run, or array to set
-     *   properties on that event
-     * @return \Zumba\Amplitude\Event
-     */
-    public function event($event = null)
+    public function newEvent(string $eventType, array $eventProperties = [])
```
Comment (jonyo) [inline src/Amplitude.php:157]: 👍 to adding type declarations, but if we add here we should be consistent and add throughout the library (or at the very least, all the methods on this class).  For instance, the constructor, the `getInstance`, `init` do not have any (just to name a few).

File: src/Amplitude.php [hunk 5]
```diff
@@ -179,39 +154,34 @@ public function resetQueue()
...
-     * Gets the event that will be used for the next event logged by call to logEvent() or queueEvent()
-     *
-     * You can also pass in an event or array of event properties.  If you pass in an event, it will be set as the
-     * event to be used for the next call to queueEvent() or logEvent()
-     *
-     * @param null|array|\Zumba\Amplitude\Event Can pass in an event to set as the next event to run, or array to set
-     *   properties on that event
-     * @return \Zumba\Amplitude\Event
-     */
-    public function event($event = null)
```
Comment (jonyo) [inline src/Amplitude.php:192]: This is pretty significantly changing how the library works, a large portion of the readme will also need to be rewritten to properly document how it works now.  I forget the reason the library held onto an event, maybe to mimic how the official JS library worked at the time or something.  Or maybe to make it easy for different parts of the code to add to an event before it is sent off.  Anyways, whatever the original reason was, if we do change that, the readme needs to be updated.

For instance, one section:

https://github.com/zumba/amplitude-php/blob/master/README.md#adding-user-properties-on-event-object

I'm OK with the change, as long as it does not remove functionality.  For instance, instead of getting a new event by calling `->event()`, you just create the event then pass it around, that is fine.  It is probably good to avoid single getter/setter methods (separation of concerns), if we made this library from scratch today we would probably do something similar.

But as I said the readme will need to be re-written to update the documentation.  Can't have a bunch of non-working example code. 😄

File: src/Amplitude.php [hunk 6]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
+     * @param resource $ch
+     * @throws \Exception
+     */
+    protected function checkCurlHandle($ch): void
+    {
+        if (!$ch) {
+            // Could be a number of PHP environment problems, log a critical error
+            $message = 'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event';
+            $this->logger->critical($message);
+            throw new \Exception($message);
```
Comment (jonyo) [inline src/Amplitude.php:518]: The current methodology of this library is to throw exceptions for anything that is a code related error (things that need to be caught hopefully before they make it to production), and log everything else.  At the end of the day this is a library for logging data for analytics on production environments.

Another way to put it, if amplitude goes down, do you want your own app to stop working completely?  Because without catching this exception, that is what would happen.

That said, using exceptions does make it easier to exit early and makes for much cleaner code in the library.  If we do change it however, we need to re-write the docs to make it clear the things that could possibly throw an exception, and make it clear that in a production env they should always wrap it in a try/catch block to prevent an outage just because amplitude goes out or something.

**tl;dr:** If we start using exceptions for things that are not code related, we need to update the documentation in the readme as well.

File: src/Amplitude.php [hunk 7]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
+        if (empty($this->apiKey)) {
+            throw new \LogicException(static::EXCEPTION_MSG_NO_API_KEY);
+        }
+    }
+
+    /**
+     * @param resource $ch
+     * @throws \Exception
+     */
+    protected function checkCurlHandle($ch): void
```
Comment (jonyo) [inline src/Amplitude.php:512]: If you use return declarations, need to do it consistently, either on all methods or none.  No harm in adding them to all the methods since it will already require a major version bump with all the non-BC changes.

File: src/Amplitude.php [hunk 8]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
 
+    /**
+     * Use the Identify API to set the User ID for a particular Device ID or update user properties
+     * of a particular user without sending an event. You can modify Amplitude default user
+     * properties as well as custom user properties that you have defined. However, these updates
+     * will only affect events going forward.
+     *
+     * Note that api key, and either the user ID or device ID need to be set prior to calling this.
+     */
+    public function identify()
```
Comment (jonyo) [inline src/Amplitude.php:407]: For consistency, this should return `$this` to allow chaining, you already are if the `optOut` flag is thrown...

File: src/Amplitude.php [hunk 9]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
+        if ($this->optOut) {
+            return $this;
+        }
+
+        $this->sendIdentify();
+
+        $this->resetUserProperties();
+    }
+
+    protected function sendIdentify()
```
Comment (jonyo) [inline src/Amplitude.php:418]: Needs a docblock.

File: src/Amplitude.php [hunk 10]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
      * Send the event currently set in $this->event to amplitude
      *
      * Requres $this->event and $this->apiKey to be set, otherwise it throws an exception.
      *
      * @return void
-     * @throws \InternalErrorException If event or api key not set
+     * @throws AmplitudeException If event or api key not set
      */
-    protected function sendEvent()
+    protected function sendEvent(Event $event)
```
Comment (jonyo) [inline src/Amplitude.php:464]: In addition to updating the readme as I mentioned in another comment... Here, need to update the docblock.  It currently says that this method will send the event currently set in `$this->event` but that is no longer a thing in this PR...

Comment (jonyo) [reply src/Amplitude.php:464]: It would probably be good to check all the docblocks to update references throughout the file.

I know the old saying is...

> Code never lies, but comments do

But lets try to make the comments as truthful as possible  😄

File: src/Amplitude.php [hunk 11]
```diff
@@ -474,49 +396,167 @@ public function getOptOut()
...
+     * @throws AmplitudeException
+     */
+    protected function checkForCurlError($ch): void
+    {
+        if ($curlErrno = curl_errno($ch)) {
+            $message = 'Curl error: ' . curl_error($ch);
+            $context = compact('curlErrno', 'response', 'postFields');
+            $this->logger->critical($message, $context);
+            curl_close($ch);
+            throw new \Exception($message);
```
Comment (jonyo) [inline src/Amplitude.php:533]: Same here, either need to update the readme, or keep the current philosophy of "exceptions for code related errors, log for connection problems or anything else not due to code".

**Review by jonyo — CHANGES_REQUESTED** [review]
In case you did not notice, you have failing tests, looks like linter issues.

edit: Also I meant to say, thanks for taking the time to put this PR together!

**Comment by jonyo** [comment]
@lunfel What do you think of breaking up this PR into smaller PR's?  For instance, one to add the identify stuff, one to update the API endpoint (and anything needed to do that), one to update PHPUnit, etc.

As it is this is just a monster PR and is part of the reason it took a long time to find the time to go over it completely.  If you break it up, each part will be easier to review.  It would also let us give each part the attention it needs and less chance of something getting overlooked.

And we could hold off on making a new release until all of the changes are merged in.

**Comment by lunfel** [comment]
> @lunfel What do you think of breaking up this PR into smaller PR's? For instance, one to add the identify stuff, one to update the API endpoint (and anything needed to do that), one to update PHPUnit, etc.
> 
> As it is this is just a monster PR and is part of the reason it took a long time to find the time to go over it completely. If you break it up, each part will be easier to review. It would also let us give each part the attention it needs and less chance of something getting overlooked.
> 
> And we could hold off on making a new release until all of the changes are merged in.

Yes, I could definitively break it down into smaller parts. I don't know when I'll have time to do it though.

Thanks for taking time to review my PR. All your feedback makes sense.

File: src/Amplitude.php (hunk 11 above)
Comment (jonyo) [reply src/Amplitude.php:533]: For posterity, I just realized the "same here" is referring to a comment I ended up removing...  The original comment was basically what I commented on #16 that currently the design is that it only throws exceptions for "coding errors" not "connection issues".  

While that is how it is currently designed, if changing that allows for much cleaner / easier to maintain code in the library, I think it would be fine to can change that.  As long as we also update the docs to make it very clear you always need a try/catch around any amplitude calls to avoid possible downtime yourself.

**Comment by RebornGeek** [comment]
Any chance we can get this PR updated and merged in? HTTP v2 would be preferable to use at this point

**Comment by lunfel** [comment]
@RebornGeek I unfortunately don't have any free time to adjust the PR with the feedback. But you are welcome to make a new PR using my code.
